File: 프로그래머스/unrated/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    result = []
    stack = []
    for i in range(len(numbers)-1, -1, -1):
        if not stack:
            result.append(-1)
            stack.append(numbers[i])
        elif stack[-1] <= numbers[i]:
            while True:
                if not stack:
                    result.append(-1)
                    stack.append(numbers[i])
                    break
                if stack[-1] <= numbers[i]:
                    stack.pop()
                elif stack[-1] > numbers[i]:
                    result.append(stack[-1])
                    stack.append(numbers[i])
                    break
        elif stack[-1] > numbers[i]:
            result.append(stack[-1])
            stack.append(numbers[i])

    return result[::-1]

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        Stack<Integer> stack = new Stack<>();

        Arrays.fill(answer, -1);

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
                answer[stack.pop()] = numbers[i];
            }
            stack.push(i);
        }

        return answer;
    }
}